working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+          mainAxisSize: MainAxisSize.min,
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [Text('SHARE')],
+        ),
+      ],
+    );
+  }
+}
+```
+
+Or this:
+
+```dart
+class MyApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      children: [
+        _buildColumn('CALL'),
+        _buildColumn('ROUTE'),
+        _buildColumn('SHARE'),
+      ],
+    );
+  }
+
+  Column _buildColumn(String label) {
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [Text(label)],
+    );
+  }
+}
+```
+
+Loops and functions are powerful, but fundamentally limited. They only execute
+within imperative code—initializers and function bodies—and they can only
+operate on data—values that are first-class in Dart. If you want to, say,
+iterate over all the fields in a class (maybe to serialize them), you have to
+use the “dart:mirrors” library which exists to expose data about classes as
+first-class Dart objects. That works, but sacrifices performance, causes code
+size problems, and isn’t supported at all in AOT builds.
+
+In the above example, we can hoist the repetitive code to create a column into a
+function because Flutter widgets are built using imperative code inside
+`build()` methods. But say you wanted to define each column as a separate widget
+class, like this:
+
+```dart
+class CallBuildColumn extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [Text('CALL')],
+    );
+  }
+}
+
+class RouteBuildColumn extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [Text('ROUTE')],
+    );
+  }
+}
+
+class ShareBuildColumn extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [Text('SHARE')],
+    );
+  }
+}
+```
+
+There is no function you can write in Dart and call three times to output these
+three classes, short of a heavy code generator. Dart doesn’t let you abstract
+over classes or other top-level declarations.
+
+**Metaprogramming** refers to code that can do this—code that operates on other
+code as if it were data. It can take code in as parameters, reflect over it,
+inspect it, create it, modify it, and return it. ***Static* metaprogramming**
+means doing that work at compile-time. This avoids the safety, performance, and
+code size problems of runtime metaprogramming. Examples are the C preprocessor,
+C++ templates, Rust macros, Swift [function builders][function_builders], and
+[compile-time execution in Zig][zig_compile_time_execution].
```

Comment:
and generics in Dart, to a very small limited extent.

Reply:
D popularized CTFE quite some time before Zig, by the way: https://dl.acm.org/doi/pdf/10.1145/3386323